infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -173,226 +182,265 @@ public void handleTailRequest(CorfuPayloadMsg<TailsRequest> msg, ChannelHandlerC
                     return null;
                 });
     }
 
     /**
      * Service an incoming request to retrieve the starting address of this logging unit.
      */
-    @ServerHandler(type = CorfuMsgType.TRIM_MARK_REQUEST)
-    public void handleTrimMarkRequest(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r) {
-        log.trace("handleTrimMarkRequest: received a trim mark request {}", msg);
-        r.sendResponse(ctx, msg, CorfuMsgType.TRIM_MARK_RESPONSE.payloadMsg(streamLog.getTrimMark()));
+    @RequestHandler(type = PayloadCase.TRIM_MARK_REQUEST)
+    public void handleTrimMarkRequest(RequestMsg req, ChannelHandlerContext ctx, IServerRouter r) {
+        if (log.isTraceEnabled()) {
+            log.trace("handleTrimMarkRequest: received a trim mark request {}", TextFormat.shortDebugString(req));
+        }
+
+        // Note: we reuse the request header as the ignore_cluster_id and
+        // ignore_epoch fields are the same in both cases.
+        r.sendResponse(getResponseMsg(req.getHeader(), getTrimMarkResponseMsg(streamLog.getTrimMark())), ctx);
     }
 
     /**
      * Service an incoming query for the committed tail on this log unit server.
      */
-    @ServerHandler(type = CorfuMsgType.COMMITTED_TAIL_REQUEST)
-    public void handleCommittedTailRequest(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r) {
-        log.trace("handleCommittedTailRequest: received a committed log tail request {}", msg);
-        r.sendResponse(ctx, msg, CorfuMsgType.COMMITTED_TAIL_RESPONSE.payloadMsg(streamLog.getCommittedTail()));
+    @RequestHandler(type = PayloadCase.COMMITTED_TAIL_REQUEST)
+    public void handleCommittedTailRequest(RequestMsg req, ChannelHandlerContext ctx, IServerRouter r) {
+        if (log.isTraceEnabled()) {
+            log.trace("handleCommittedTailRequest: received a "
+                    + "committed log tail request {}", TextFormat.shortDebugString(req));
+        }
+
+        // Note: we reuse the request header as the ignore_cluster_id and
+        // ignore_epoch fields are the same in both cases.
+        r.sendResponse(getResponseMsg(req.getHeader(),
+                getCommittedTailResponseMsg(streamLog.getCommittedTail())), ctx);
     }
 
     /**
      * Service an incoming request to update the current committed tail.
      */
-    @ServerHandler(type = CorfuMsgType.UPDATE_COMMITTED_TAIL)
-    public void updateCommittedTail(CorfuPayloadMsg<Long> msg,
-                                    ChannelHandlerContext ctx, IServerRouter r) {
-        log.trace("updateCommittedTail: received request to update committed tail {}", msg);
-        streamLog.updateCommittedTail(msg.getPayload());
-        r.sendResponse(ctx, msg, CorfuMsgType.ACK.msg());
+    @RequestHandler(type = PayloadCase.UPDATE_COMMITTED_TAIL_REQUEST)
+    public void handleUpdateCommittedTailRequest(RequestMsg req, ChannelHandlerContext ctx, IServerRouter r) {
+        if (log.isTraceEnabled()) {
```

Comment:
Instead of introducing an if statement, you can simply emulate lazy-evaluation:

```
    interface LazyString {
        String toString();

        static LazyString create(MessageOrBuilder message) {
            return new LazyString() {
                @Override
                public String toString() {
                    return TextFormat.shortDebugString(message);
                }
            };
        }
    }
```
Which will enable you to do:

```
log.trace("handleLogAddressSpaceRequest[{}]: received a log " +
        "address space request {}", req.getHeader().getRequestId(), LazyString.create(req));
```

Alternatively, we can upgrade `slf4j` to include the fluent API, which would allow you to pass a `Supplier`. However, that is a riskier change.

Reply:
Although this emulates lazy-evaluation, it does not guard against the allocation of an additional object like the if statement would when `trace`/`debug` is disabled. I thinking keeping the if statement is slightly better since it provides both benefits.